### PR TITLE
Partially support DetectionOutput layer

### PR DESCRIPTION
**Pull request by fantasyRqg**
Partially support DetectionOutput layer

### CODE REVIEW

File: mace/ops/detection_output.cc [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+  int keep_top_k_;
+  float confidence_threshold_;
+};
+
+
+void RegisterDetectionOutput(OpRegistry *op_registry) {
+  MACE_REGISTER_OP(op_registry, "DetectionOutput", DetectionOutputOp,
+                   RuntimeType::RT_CPU, float);
+
+  MACE_REGISTER_OP_CONDITION(
```
Comment (lu229) [inline mace/ops/detection_output.cc:252]: If you did not register GPU OP, MACE_REGISTER_OP_CONDITION is not needed.

File: tools/python/transform/caffe_converter.py [hunk 2]
```diff
@@ -635,6 +634,41 @@ def convert_concat(self, caffe_op):
...
+
+        nms_threshold = op.arg.add()
+        nms_threshold.f = param.nms_param.nms_threshold
+        nms_threshold.name = MaceKeyword.mace_nms_threshold
+
+        top_k = op.arg.add()
+        top_k.i = param.nms_param.top_k
+        top_k.name = MaceKeyword.mace_nms_top_k
+
+        if param.code_type != 2:
```
Comment (lu229) [inline tools/python/transform/caffe_converter.py:660]: mace_check

File: tools/python/transform/shape_inference.py [hunk 3]
```diff
@@ -210,7 +211,14 @@ def infer_shape_concat(self, op):
             input_shape = list(self._output_shape_cache[input_node])
             output_shape[axis] = output_shape[axis] + input_shape[axis]
         self.add_output_shape(op, [output_shape])
-
```
Comment (lu229) [inline tools/python/transform/shape_inference.py:213]: add an empty line

File: tools/python/transform/shape_inference.py [hunk 4]
```diff
@@ -210,7 +211,14 @@ def infer_shape_concat(self, op):
...
+    def infer_shape_detection_output(self,op):
+        keep_top = 0
+        for arg in op.arg:
+            if arg.name == MaceKeyword.mace_keep_top_k:
+                keep_top = arg.i
+                break
+        outshape = [1,1,keep_top,7]
+        self.add_output_shape(op,[outshape])
     def infer_shape_slice(self, op):
```
Comment (lu229) [inline tools/python/transform/shape_inference.py:222]: add an empty line

File: mace/ops/detection_output.cc [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+    const float *conf_ptr = conf_t->data<float>();
+    const float *pbox_ptr = pbox_t->data<float>();
+
+    std::vector<BBox> bbox_rects;
+
+    DetectionOutput_CLA(loc_ptr, conf_ptr, pbox_ptr, num_prior, num_classes_,
+                        nms_threshold_, nms_top_k_, keep_top_k_,
+                        confidence_threshold_, &bbox_rects);
+
+    output->Clear();
```
Comment (lu229) [inline mace/ops/detection_output.cc:219]: Can output > clear() be deleted?

File: tools/python/transform/caffe_converter.py [hunk 6]
```diff
@@ -12,7 +12,6 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-
```
Comment (lu229) [inline tools/python/transform/caffe_converter.py:15]: No need to delete?

File: tools/python/transform/caffe_converter.py [hunk 7]
```diff
@@ -146,7 +145,6 @@ def add_blob(self, weight):
             op = self._ops[weight.name]
             op.blobs = list(weight.blobs)
 
-
```
Comment (lu229) [inline tools/python/transform/caffe_converter.py:149]: No need to delete?

File: tools/python/transform/caffe_converter.py [hunk 8]
```diff
@@ -635,6 +634,41 @@ def convert_concat(self, caffe_op):
...
+    def convert_detection_output(self, caffe_op):
+        op = self.convert_general_op(caffe_op)
+        param = caffe_op.layer.detection_output_param
+        op.type = MaceOp.DetectionOutput.name
+
+        num_classes = op.arg.add()
+        num_classes.name = MaceKeyword.mace_num_classes
+        num_classes.i = param.num_classes
+
+        if not param.share_location:
```
Comment (lu229) [inline tools/python/transform/caffe_converter.py:646]: mace_check

File: tools/python/transform/caffe_converter.py [hunk 9]
```diff
@@ -635,6 +634,41 @@ def convert_concat(self, caffe_op):
...
+            raise Exception('detection output only supports share location')
+
+        if param.background_label_id != 0:
```
Comment (lu229) [inline tools/python/transform/caffe_converter.py:649]: mace_check

File: mace/ops/detection_output.cc [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+            Operation::GetOptionalArg<float>("confidence_threshold", 0.05f)) {}
+
+  MaceStatus Run(OpContext *context) override {
+    MACE_UNUSED(context);
+    Tensor *output = this->Output(0);
+
+    auto *loc_t = this->Input(0);
+    auto *conf_t = this->Input(1);
+    auto *pbox_t = this->Input(2);
+
```
Comment (lu229) [inline mace/ops/detection_output.cc:200]: Delete an empty line

File: mace/ops/detection_output.cc [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+                  bboxes[2 + j * 4],
+                  bboxes[3 + j * 4],
+                  i,
+                  confidence};
+        class_bbox_rects.push_back(c);
+      }
+    }
+    std::sort(class_bbox_rects.begin(), class_bbox_rects.end(), cmp);
+
+    // apply nms
```
Comment (lu229) [inline mace/ops/detection_output.cc:146]: Apply nms

File: mace/ops/detection_output.cc [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+    std::vector<BBox> sorted_boxes;
+    NmsSortedBboxes(class_bbox_rects, nms_threshold,
+                    std::min(top_k, static_cast<int>(class_bbox_rects.size())),
+                    &sorted_boxes);
+    // gather
```
Comment (lu229) [inline mace/ops/detection_output.cc:151]: Gather

File: mace/ops/detection_output.cc [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+
+    bboxes[0 + index] = bbox_cx - bbox_w * 0.5f;
+    bboxes[1 + index] = bbox_cy - bbox_h * 0.5f;
+    bboxes[2 + index] = bbox_cx + bbox_w * 0.5f;
+    bboxes[3 + index] = bbox_cy + bbox_h * 0.5f;
+  }
+  // start from 1 to ignore background class
+
+  for (int i = 1; i < num_classes; ++i) {
+    // filter by confidence threshold
```
Comment (lu229) [inline mace/ops/detection_output.cc:130]: Filter xxx

File: mace/ops/detection_output.cc [hunk 14]
```diff
@@ -0,0 +1,262 @@
...
+    float bbox_cy = var[1] * lc[1] * pb_h + pb_cy;
+    float bbox_w = std::exp(var[2] * lc[2]) * pb_w;
+    float bbox_h = std::exp(var[3] * lc[3]) * pb_h;
+
+    bboxes[0 + index] = bbox_cx - bbox_w * 0.5f;
+    bboxes[1 + index] = bbox_cy - bbox_h * 0.5f;
+    bboxes[2 + index] = bbox_cx + bbox_w * 0.5f;
+    bboxes[3 + index] = bbox_cy + bbox_h * 0.5f;
+  }
+  // start from 1 to ignore background class
```
Comment (lu229) [inline mace/ops/detection_output.cc:127]: Start from xxx

File: mace/ops/detection_output.cc [hunk 15]
```diff
@@ -0,0 +1,262 @@
+// Copyright 2018 The MACE Authors. All Rights Reserved.
```
Comment (lu229) [inline mace/ops/detection_output.cc:1]: 2021

**Comment by lu229** [comment]
Thank you for your contribution, which is an excellent submission. I'm sorry for the late code review.

File: mace/ops/detection_output.cc (hunk 5 above)
Comment (fantasyRqg) [reply mace/ops/detection_output.cc:219]: No, necessory for clear previous inference output. `bbox_rects.size()` is not fixed

Comment (lu229) [reply mace/ops/detection_output.cc:218]: I see 'output > Reset (output)' on line 220, If the shape of output is correct, the output data will cover the original data and the redundant data will not be accessed?

**Comment by icyhearts** [comment]
Would you please run tools/cpplint.sh and pycodestyle to find coding style errors reported by cpplint and pycodestyle, and fix them?

.gitlab-ci.yml shows the usage of these tools.

**Comment by icyhearts** [comment]
Would you please add some unit test cases for new added Op?
There are examples in test/ccunit/mace/ops/.

**Comment by icyhearts** [comment]
1. Please run 
```
bash tools/bazel_build_standalone_lib.sh  -abi=arm64-v8a --runtimes=gpu --static
```
to make sure that the compiler will not report errors.

2. Please run you test cases by this command:
```
python tools/bazel_adb_run.py --target="//test/ccunit:mace_cc_test" --run_target=True --stdout_processor=unittest_stdout_processor --target_abis=arm64-v8a
```
Make sure that the compiler will not report errors and  your new added test cases pass.
It seems that you added test cases in ```test/ccunit/mace/ops/detection_output_test.cc```, but running it was not done.
For example, I run 
```
python tools/bazel_adb_run.py --target="//test/ccunit:mace_cc_test" --run_target=True --stdout_processor=unittest_stdout_processor --target_abis=arm64-v8a
```
But  got these errors:
```
test/ccunit/mace/ops/detection_output_test.cc:32:36: error: no template named 'vcetor' in namespace 'std'; did you mean 'vector'?
                        const std::vcetor<int> &offset,
                              ~~~~~^~~~~~
                                   vector
external/androidndk/ndk/sources/cxx-stl/llvm-libc++/include/iterator:1407:74: note: 'vector' declared here
    template <class _Tp, class _Alloc> friend class _LIBCPP_TEMPLATE_VIS vector;
                                                                         ^
test/ccunit/mace/ops/detection_output_test.cc:32:49: error: unused parameter 'offset' [-Werror,-Wunused-parameter]
                        const std::vcetor<int> &offset,
                                                ^
test/ccunit/mace/ops/detection_output_test.cc:71:3: error: no matching function for call to 'RunDetectionOutput'
  RunDetectionOutput<RuntimeType::RT_CPU>(
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
test/ccunit/mace/ops/detection_output_test.cc:26:6: note: candidate function template not viable: requires 14 arguments, but 13 were provided
void RunDetectionOutput(const std::vector<index_t> &input_shape1,
     ^
3 errors generated.

```

**Comment by fantasyRqg** [comment]
Test Env setup is too complicate. I tried MacOsX and Ubuntu. MacOsX built failed, Ubuntu can't find my Android devcie.

**Comment by fantasyRqg** [comment]
```
➜  mace git:(master) bash tools/bazel_build_standalone_lib.sh  -abi=arm64-v8a --runtimes=gpu --static

-------------build static lib for arm64-v8a +  gpu------------
WARNING: The major revision of the Android NDK referenced by android_ndk_repository rule 'androidndk' is 19. The major revisions supported by Bazel are [10, 11, 12, 13, 14, 15, 16]. Bazel will attempt to treat the NDK as if it was r16. This may cause compilation and linkage problems. Please download a supported NDK version.
ERROR: /Users/rqg/Playground/mace/mace/codegen/BUILD.bazel:26:1: no such target '@local_opencl_kernel_encrypt//:gen/encrypt_opencl_kernel.cc': target 'gen/encrypt_opencl_kernel.cc' not declared in package '' defined by /private/var/tmp/_bazel_rqg/a266a722dbdbd6fad8042544d6cf7b92/external/local_opencl_kernel_encrypt/BUILD.bazel and referenced by '//mace/codegen:encrypt_opencl_kernel_gen'
ERROR: /Users/rqg/Playground/mace/mace/codegen/BUILD.bazel:26:1: no such target '@local_opencl_kernel_encrypt//:gen/encrypt_opencl_kernel.h': target 'gen/encrypt_opencl_kernel.h' not declared in package '' defined by /private/var/tmp/_bazel_rqg/a266a722dbdbd6fad8042544d6cf7b92/external/local_opencl_kernel_encrypt/BUILD.bazel and referenced by '//mace/codegen:encrypt_opencl_kernel_gen'
ERROR: Analysis of target '//mace/libmace:libmace_static' failed; build aborted: Loading failed
INFO: Elapsed time: 0.772s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (3 packages loaded)
```

**Comment by icyhearts** [comment]
If it's hard to make unit tests pass, you can remove your added unit tests.
If the environment is hard to prepare, you can use docker: https://mace.readthedocs.io/en/latest/installation/using_docker.html
But correct compilation without errors is required.

**Pull request merged**